pkg/controller/tas/node_failure_controller_test.go

```diff
@@ -0,0 +1,325 @@
+	cl, r := setupTest(fakeClock,
+		newNodeTest(nodeName, corev1.ConditionTrue),
+		baseWorkload.DeepCopy(),
+		basePod.DeepCopy(),
+	)
+
+	t.Log("Step 1: Reconciling healthy node")
+	result, err := r.Reconcile(ctx, reconcile.Request{NamespacedName: types.NamespacedName{Name: nodeName}})
+	if err != nil {
+		t.Fatalf("Reconcile healthy node failed: %v", err)
+	}
+	wl := &kueue.Workload{}
+	if err := cl.Get(ctx, wlKey, wl); err != nil {
+		t.Fatalf("Failed to get workload: %v", err)
+	}
+
+	annotations := wl.GetAnnotations()
+	if ann, ok := annotations[kueuealpha.NodeToReplaceAnnotation]; ok && ann != "" {
+		actualFailedNodes := strings.Split(ann, ",")
+		if len(actualFailedNodes) > 0 && (len(actualFailedNodes) > 1 || actualFailedNodes[0] != "") {
+			t.Errorf("Expected FailedNodes annotation to be empty or not present, got %v", ann)
+		}
+	}
+	if result.RequeueAfter != 0 {
+		t.Errorf("Expected no requeue for healthy node, got %v", result.RequeueAfter)
+	}
+
+	t.Log("Step 2: Simulating node becoming NotReady")
+	setNodeConditionTest(ctx, t, cl, fakeClock, nodeName, corev1.ConditionFalse)
+
+	t.Log("Step 3: Reconciling unhealthy node")
+	result, err = r.Reconcile(ctx, reconcile.Request{NamespacedName: types.NamespacedName{Name: nodeName}})
+	if err != nil {
+		t.Fatalf("Reconcile unhealthy node (before delay) failed: %v", err)
+	}
+	if err := cl.Get(ctx, wlKey, wl); err != nil {
+		t.Fatalf("Failed to get workload: %v", err)
+	}
+	annotations = wl.GetAnnotations()
+	if ann, ok := annotations[kueuealpha.NodeToReplaceAnnotation]; ok && ann != "" {
+		actualFailedNodes := strings.Split(ann, ",")
+		if len(actualFailedNodes) > 0 && (len(actualFailedNodes) > 1 || actualFailedNodes[0] != "") {
+			t.Errorf("Expected FailedNodes annotation to be empty or not present before delay, got %v", ann)
+		}
+	}
+	if result.RequeueAfter <= 0 {
+		t.Errorf("Expected requeue after delay, got %v", result.RequeueAfter)
+	}
+
+	t.Log("Step 4: Advancing clock past delay")
+	fakeClock.Step(NodeFailureDelay + time.Second)
+
+	t.Log("Step 5: Reconciling unhealthy node after delay")
+	result, err = r.Reconcile(ctx, reconcile.Request{NamespacedName: types.NamespacedName{Name: nodeName}})
+	if err != nil {
+		t.Fatalf("Reconcile unhealthy node failed: %v", err)
+	}
+	if err := cl.Get(ctx, wlKey, wl); err != nil {
+		t.Fatalf("Failed to get workload: %v", err)
+	}
+	annotations = wl.GetAnnotations()
+	annValue, found := annotations[kueuealpha.NodeToReplaceAnnotation]
+	if !found {
+		t.Fatalf("Expected annotation %q to be present", kueuealpha.NodeToReplaceAnnotation)
+	}
+	actualFailedNodes := strings.Split(annValue, ",")
+	if !slices.Contains(actualFailedNodes, nodeName) {
+		t.Errorf("Expected node %q to be in FailedNodes annotation %v, got %v", nodeName, kueuealpha.NodeToReplaceAnnotation, actualFailedNodes)
+	}
+	if result.RequeueAfter != 0 {
+		t.Errorf("Expected no requeue after marking failed, got %v", result.RequeueAfter)
+	}
+}
```

Comment:
Is there any case here that is not reflected in the scenarios above? If not, I suggest to keep tests as simple as possible and remove that. For the whole lifecycle we can have integration/e2e tests